### PR TITLE
ddl: support `admin cancel ddl job` when the DDL worker has started the job

### LINKED ISSUE #17904: `ADMIN CANCEL DDL JOB` doesn't take effect when the DDL worker has started the job
### Background

The main logic of cancelling a DDL job is:
1.1. Fetch the job from a DDL job queue.
1.2. If the job can be cancelled, modify its state to `CANCELLING`.
1.3. Store the job back into the job queue.

When a TiDB DDL worker runs a DDL job, the main logic is:
2.1. Fetch the first job from a DDL job queue.
2.2. If the job state is `CANCELLING`, rollback the job and repeat from 2.1.
2.3. Do the specific DDL work, e.g. create table, add index.
2.4. Update the job state and store it back to the job queue.
2.5. Repeat from 2.1 to 2.4.

### Problems and solutions

When a user cancels a job by `ADMIN CANCEL DDL JOB` while the DDL worker has arrived at step 2.3, there are 2 problems here.

Problem: If the job has finished step 2.3 successfully, the job state will be overridden as `DONE` at step 2.4. Although the cancelling action has been successfully executed, it hasn't been perceived by the DDL worker immediately.
Solution: When the DDL worker arrives at step 2.4, it checks the latest job state from the queue firstly. If the latest state is changed to `CANCELLING`, it rollbacks the job.

Problem: If the job is blocked at step 2.3, which is the most common case for cancelling, the worker won't percept the cancelling state of the job and stop it immediately.
Soltuion: A background thread periodically fetches the latest job state from the job queue to check whether it has been cancelled. The DDL worker reads the state and decides to continue or not.

### Optimization

As fetching the entire job is quite expensive, we can optimize it with maintaining a cancelled job-id list in TiKV.  The progress is:
1. `ADMIN CANCEL DDL JOB` pushes the job id into the list.
2. The background thread / worker only fetches the job id from the list.
3. When the job is rollbacking, pop the job id from the list.

## Score
- 3000

## SIG slack channel
- channel #sig-ddl in the [tidbcommunity](https://join.slack.com/t/tidbcommunity/shared_invite/zt-9vpzdqh2-8LsybcK0US_nqwvfAjSU5A) slack workspace, you can join this channel through [this invitation link](https://slack.tidb.io/invite?team=tidb-community&channel=sig-ddl&ref=pingcap-community).

## Mentor
- AilinKid
- Zimulala

## Recommended Skills

[TiDB DDL architecture](https://github.com/pingcap/tidb/blob/master/docs/design/2018-10-08-online-DDL.md)

### PR DESCRIPTION
**Pull request by TszKitLo40**
### What problem does this PR solve?

Issue Number: close #17904 
Problem Summary:

### What is changed and how it works?

What's Changed:
DDL worker check the latest state of job before update the job state and store it back to the job queue.
How it Works:
If the latest state of the job is `cancelling`, rollback the job.

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- Support `admin cancel ddl job` when the DDL worker has started the job

### CODE REVIEW

File: ddl/ddl_worker.go [hunk 1]
```diff
@@ -527,6 +535,42 @@ func (w *worker) handleDDLJobQueue(d *ddlCtx) error {
...
+		if canceledJob.ID == job.ID {
+			found = true
+			pos = i
+			break
+		}
+	}
+	if !found {
+		return nil
+	}
+	cancelledJobs = append(cancelledJobs[:pos], cancelledJobs[pos+1:]...)
```
Comment (djshow832) [inline ddl/ddl_worker.go:570]: Don't you store it back?

File: ddl/ddl_worker.go [hunk 2]
```diff
@@ -527,6 +535,42 @@ func (w *worker) handleDDLJobQueue(d *ddlCtx) error {
...
+}
+
+func removeJobFromCancellingList(t *meta.Meta, job *model.Job) error {
+	cancelledJobs, err := getDDLJobsInQueue(t, meta.CancelJobListKey)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	found := false
+	pos := 0
+	for i, canceledJob := range cancelledJobs {
```
Comment (djshow832) [inline ddl/ddl_worker.go:560]:
```suggestion
	for i, cancelledJob := range cancelledJobs {
```

**Review by djshow832 — COMMENTED** [review]
Where do you check the DDLJobCancel queue and cancel them?

File: ddl/ddl_worker.go (hunk 1 above)
Comment (bb7133) [reply ddl/ddl_worker.go:587]: Is storing back required?

Comment (djshow832) [reply ddl/ddl_worker.go:587]: If not, why assign a value to a local variable at the end of this function?
Every time it checks the cancel queue, it reads from metadata, not memory.

Comment (TszKitLo40) [reply ddl/ddl_worker.go:587]: This function will be called `finishDDLJob`. In `finishDDLJob`, I think it should be removed from the queue and there is no need to store it back.

File: ddl/ddl_worker.go [hunk 3]
```diff
@@ -527,6 +539,55 @@ func (w *worker) handleDDLJobQueue(d *ddlCtx) error {
 	}
 }
 
+func getDDLJobsInQueue(t *meta.Meta, jobListKey meta.JobListKeyType) ([]*model.Job, error) {
```
Comment (bb7133) [inline ddl/ddl_worker.go:542]: Looks that the code is copied from: https://github.com/pingcap/tidb/blob/f121cb7682894e88cfc58380924c10de99e7ed20/util/admin/admin.go#L200, right?

I think we can keep just one of them, and export it to be re-used.

Comment (TszKitLo40) [reply ddl/ddl_worker.go:542]: Yes. But `getDDLJobsInQueue` in `admin.go` is private. Do you think can we make it public to other package?

File: ddl/ddl_worker.go (hunk 1 above)
Comment (bb7133) [reply ddl/ddl_worker.go:587]: I see, we need to store it back(the 'metadata' in TiKV) or the persisted queue will be increasing still.

Comment (TszKitLo40) [reply ddl/ddl_worker.go:587]: When should we remove the job from the `cancelling queue`?

File: ddl/ddl_worker.go (hunk 3 above)
Comment (AilinKid) [inline ddl/ddl_worker.go:542]: Maintaining a job list rather than a job id list is quite heavy, doesn't it?

File: ddl/ddl_worker.go (hunk 1 above)
Comment (AilinKid) [reply ddl/ddl_worker.go:587]: We may need to store it back once after we did the check here. otherwise, the rolling back logic will get in a loop.

**Review by AilinKid — COMMENTED** [review]
Sorry for the postponed review.

File: ddl/ddl_worker.go (hunk 1 above)
Comment (TszKitLo40) [reply ddl/ddl_worker.go:587]: But this function is called in `finishDDLJob` and the job has been rollback. In `finishDDLJob`, do you think we should remove it from the queue? Or when the job should be removed from the queue?

**Comment by AilinKid** [comment]
cancel the job when the job is already in cancelling will get stuck. 
```
// Check whether a job is cancelling when we are handling it.
if inCancellingList, err := isInCancellingList(t, job); err == nil {
   if inCancellingList {
      if job.State != model.JobStateRollingback && job.State != model.JobStateCancelled && job.State != model.JobStateRollbackDone {
         // If we found a job is cancelling here, we should abandon the kv txn modification,
         // and skip storing the job modification down. the next ddl round will handle the
         // cancelling logic for it.
         txn.Reset()
         logutil.Logger(w.logCtx).Info("[ddl] DDL job kv modification abandoned, since it has been cancelling.")
         err := removeJobFromCancellingList(t, job)
         if err != nil {
            return errors.Trace(err)
         }
         // return directly to step into the cancelling in next ddl round.
         return nil
      }
      // Since we are already in the cancel handling logic, keep going on.
      err := removeJobFromCancellingList(t, job)
      if err != nil {
         return errors.Trace(err)
      }
   }
}
```

**Comment by TszKitLo40** [comment]
> cancel the job when the job is already in canceling will get stuck.
> Better exclude some job state out when we are handling the cancelling state.
> Although the job state is changed to cancelling by outside, the job will be overwritten by memory state such `JobStateRollingback, JobStateCancelled, JobStateRollbackDone` in updateDDLJob.
> 
> ```
> // Check whether a job is cancelling when we are handling it.
> if inCancellingList, err := isInCancellingList(t, job); err == nil {
>    if inCancellingList {
>       if job.State != model.JobStateRollingback && job.State != model.JobStateCancelled && job.State != model.JobStateRollbackDone {
>          // If we found a job is cancelling here, we should abandon the kv txn modification,
>          // and skip storing the job modification down. the next ddl round will handle the
>          // cancelling logic for it.
>          txn.Reset()
>          logutil.Logger(w.logCtx).Info("[ddl] DDL job kv modification abandoned, since it has been cancelling.")
>          err := removeJobFromCancellingList(t, job)
>          if err != nil {
>             return errors.Trace(err)
>          }
>          // return directly to step into the cancelling in next ddl round.
>          return nil
>       }
>       // Since we are already in the cancel handling logic, keep going on.
>       err := removeJobFromCancellingList(t, job)
>       if err != nil {
>          return errors.Trace(err)
>       }
>    }
> }
> ```

**Comment by TszKitLo40** [comment]
> cancel the job when the job is already in canceling will get stuck.
> Better exclude some job state out when we are handling the cancelling state.
> Although the job state is changed to cancelling by outside, the job will be overwritten by memory state such `JobStateRollingback, JobStateCancelled, JobStateRollbackDone` in updateDDLJob.
> 
> ```
> // Check whether a job is cancelling when we are handling it.
> if inCancellingList, err := isInCancellingList(t, job); err == nil {
>    if inCancellingList {
>       if job.State != model.JobStateRollingback && job.State != model.JobStateCancelled && job.State != model.JobStateRollbackDone {
>          // If we found a job is cancelling here, we should abandon the kv txn modification,
>          // and skip storing the job modification down. the next ddl round will handle the
>          // cancelling logic for it.
>          txn.Reset()
>          logutil.Logger(w.logCtx).Info("[ddl] DDL job kv modification abandoned, since it has been cancelling.")
>          err := removeJobFromCancellingList(t, job)
>          if err != nil {
>             return errors.Trace(err)
>          }
>          // return directly to step into the cancelling in next ddl round.
>          return nil
>       }
>       // Since we are already in the cancel handling logic, keep going on.
>       err := removeJobFromCancellingList(t, job)
>       if err != nil {
>          return errors.Trace(err)
>       }
>    }
> }
> ```

Can we remove the second `removeJobFromCancellingList`?

**Comment by AilinKid** [comment]
> > cancel the job when the job is already in canceling will get stuck.
> > Better exclude some job state out when we are handling the cancelling state.
> > Although the job state is changed to cancelling by outside, the job will be overwritten by memory state such `JobStateRollingback, JobStateCancelled, JobStateRollbackDone` in updateDDLJob.
> > ```
> > // Check whether a job is cancelling when we are handling it.
> > if inCancellingList, err := isInCancellingList(t, job); err == nil {
> >    if inCancellingList {
> >       if job.State != model.JobStateRollingback && job.State != model.JobStateCancelled && job.State != model.JobStateRollbackDone {
> >          // If we found a job is cancelling here, we should abandon the kv txn modification,
> >          // and skip storing the job modification down. the next ddl round will handle the
> >          // cancelling logic for it.
> >          txn.Reset()
> >          logutil.Logger(w.logCtx).Info("[ddl] DDL job kv modification abandoned, since it has been cancelling.")
> >          err := removeJobFromCancellingList(t, job)
> >          if err != nil {
> >             return errors.Trace(err)
> >          }
> >          // return directly to step into the cancelling in next ddl round.
> >          return nil
> >       }
> >       // Since we are already in the cancel handling logic, keep going on.
> >       err := removeJobFromCancellingList(t, job)
> >       if err != nil {
> >          return errors.Trace(err)
> >       }
> >    }
> > }
> > ```
> 
> Can we remove the second `removeJobFromCancellingList`? Since we are already in the cancel handling logic, maybe it is no need to remove the job from cancelling list again because it has been removed in the first logic.

Nope, I answered in the channel.

File: util/admin/admin.go [hunk 4]
```diff
@@ -181,6 +181,11 @@ func CancelJobs(txn kv.Transaction, ids []int64) ([]error, error) {
 			} else {
 				err = t.UpdateDDLJob(int64(j), job, true)
 			}
+			if err != nil {
+				errs[i] = errors.Trace(err)
+				continue
+			}
+			err = t.AddCancellingJobList(job)
```
Comment (AilinKid) [inline util/admin/admin.go:188]: better check the duplicate there, if already in, just continue

**Review by AilinKid — COMMENTED** [review]
.

File: util/admin/admin.go (hunk 4 above)
Comment (TszKitLo40) [reply util/admin/admin.go:188]: If we check the duplicate here, it seems that the second `removeJobFromCancellingList` mentioned above can be removed.

**Comment by AilinKid** [comment]
This issue is closed because it is not like what it said.  Refer https://github.com/pingcap/tidb/issues/17904 for more details.

**Pull request closed without merging**